internal/snmp/wrapper.go

```diff
@@ -0,0 +1,173 @@
+package snmp
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/soniah/gosnmp"
+)
+
+// GosnmpWrapper wraps a *gosnmp.GoSNMP object so we can use it as a snmpConnection.
+type GosnmpWrapper struct {
+	*gosnmp.GoSNMP
+}
+
+// Host returns the value of GoSNMP.Target.
+func (gsw GosnmpWrapper) Host() string {
+	return gsw.Target
+}
+
+// Walk wraps GoSNMP.Walk() or GoSNMP.BulkWalk(), depending on whether the
+// connection is using SNMPv1 or newer.
+// Also, if any error is encountered, it will just once reconnect and try again.
+func (gsw GosnmpWrapper) Walk(oid string, fn gosnmp.WalkFunc) error {
+	var err error
+	// On error, retry once.
+	// Unfortunately we can't distinguish between an error returned by gosnmp, and one returned by the walk function.
+	for i := 0; i < 2; i++ {
+		if gsw.Version == gosnmp.Version1 {
+			err = gsw.GoSNMP.Walk(oid, fn)
+		} else {
+			err = gsw.GoSNMP.BulkWalk(oid, fn)
+		}
+		if err == nil {
+			return nil
+		}
+		if err := gsw.GoSNMP.Connect(); err != nil {
+			return fmt.Errorf("reconnecting: %w", err)
+		}
+	}
+	return err
+}
+
+// Get wraps GoSNMP.GET().
+// If any error is encountered, it will just once reconnect and try again.
+func (gsw GosnmpWrapper) Get(oids []string) (*gosnmp.SnmpPacket, error) {
+	var err error
+	var pkt *gosnmp.SnmpPacket
+	for i := 0; i < 2; i++ {
```

Comment:
I haven't seen it fail but this has been in the snmp input for a long time and I expect users are used to how it behaves so I don't want to change it with this PR.

Reply:
Sounds good. There's an open issue for one of the other plugins that retries like this, so I'd like to avoid creating more issues if we can help it, but I'm okay with solving this in a later PR.